tt_metal/api/buffer.hpp

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include "types.hpp"
+#include "tt_metal/impl/buffers/buffer.hpp"
+//==================================================
+//                  BUFFER HANDLING
+//==================================================
+
+namespace tt::tt_metal{
+namespace v1 {
+
+
+/**
+ * @brief Allocates an interleaved DRAM or L1 buffer on the device.
+ *
+ * @param config Configuration for the buffer.
+ * @return Buffer handle to the allocated buffer.
+ */
+Buffer CreateBuffer(const InterleavedBufferConfig &config);
+
+/**
+ * @brief Allocates a buffer on the device.
+ *
+ * @param buffer The buffer to allocate.
+ * @param bottom_up If true, allocates buffer from the bottom up.
+ */
+void AllocateBuffer(Buffer buffer, bool bottom_up);
+
+/**
+ * @brief Deallocates a buffer from the device.
+ *
+ * @param buffer The buffer to deallocate.
+ */
+void DeallocateBuffer(Buffer buffer);
```

Comment:
AFAIK I heard Metalium tries to copy OpenCL's API design. But OpenCL's `cl::Buffer` ,is governed by RAII. Manual management can lead to leakage, etc.. Any chance Buffer is made RAII?

Reply:
RAII imposes restrictions that require three possible constraints:
- the object is expensive to copy
- the object is move-only
- or the object is reference-counted

We want v1 handles to be objects that are cheap to copy, do not perform any potentially atomic operations like reference-counting when copied, and do not directly manage lifetime. You (or a higher-level library) can build RAII as an abstraction on top of this if desired. We prefer to provide the lowest level of abstraction, while enabling ourselves to have a flexible implementation, and using handles is one way to achieve that.